README.md
# lotus-cpr

A smart caching proxy for Lotus filecoin nodes.

## Overview

Lotus-cpr provides a read-only subset of the [Lotus](https://github.com/filecoin-project/lotus) RPC
API suitable for querying the state of the Filecoin network. Immutable blocks are cached locally
for fast retrieval using [gonudb](https://github.com/iand/gonudb) and many IO intensive RPC methods
can be served without calling Lotus. Fronting Lotus with Lotus-cpr can reduce load on the node,
reducing the chance that it falls behind when syncing with the Filecoin chain.


## Usage

Install using:

	go get -u github.com/iand/lotus-cpr

Lotus-cpr uses a multi-tier caching system. By default no caching is performed and requests are
forwarded directly to the Lotus node. 

Caching is enabled by specifying a database directory using the `--store` parameter. When set
Lotus-cpr will look for blocks in this database before fetching from upstream. Any blocks retrieved
from an upstream source will be stored in the database to satisfy future requests.

To further reduce load on the Lotus node blocks may also be retrived from a blockstore webserver
serving block data via HTTP. Use the `--blockstore-base` parameter to specify the base URL of the
blockstore. When enabled Lotus-cpr will consult the local database first, then the blockstore and
finally revert to direct node access.


                              +-----------+               +-----------+                  +-----------+
                              |           |               |           |                  |           |
    client --- json/rpc ----> | lotus-cpr | -- block? --> |   gondb   | -+--- http ----> |   Store   | 
                              |           |               |           |  |               |           |
                              +-----------+               +-----------+  |               +-----------+
                                    |                                    |
                                    |                                    |               +-----------+
                                    |                                    |               |           |
                                    +---------- proxy call --------------+- json/rpc --> |   Lotus   | 
                                                                                         |           |
                                                                                         +-----------+
                                             
The gonudb and blockstore caches only store immutable block data and Lotus-cpr will only attempt to use this data
when it is sure that the request requires no other state.

Lotus-cpr expects blockstores to make raw data for blocks available using the following URL pattern:

	{base_url}/{block_cid}/data.raw


Command line options:

 - `--api` (required) Multiaddress of Lotus node (default: "/ip4/127.0.0.1/tcp/1234/http")
 - `--api-token` (required) OAuth token for Lotus node
 - `--listen` (required) Address to start the RPC server on (default: ":33111")
 - `--store` (optional) Path to directory containing gonudb store used to cache blocks.
 - `--blockstore-baseurl` (optional) URL of http server containing blocks from the filecoin chain.


## Author

Written by:

* [Ian Davis](http://github.com/iand) - <http://iandavis.com/>

## License

This is free and unencumbered software released into the public domain. Anyone is free to 
copy, modify, publish, use, compile, sell, or distribute this software, either in source 
code form or as a compiled binary, for any purpose, commercial or non-commercial, and by 
any means. For more information, see <http://unlicense.org/> or the 
accompanying [`UNLICENSE`](UNLICENSE) file.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

These are changes that will probably be included in the next release.

### Added

 * Add command line flags for circuit breaker parameters

 
### Fixed

### Changed
 
### Removed


## [v0.2.0] - 2020-12-05

### Added

 * Record cache metrics and export for prometheus
 * Add circuit breaker between cache and lotus node

### Changed

 * Avoid writing zero sized blocks to store
 * Use block hash as key

## [v0.1.0] - 2020-11-24

Initial release


s3.go
package main

import (
	"context"
	"io/ioutil"
	"net/http"
	"path"

	"github.com/go-logr/logr"
	"github.com/ipfs/go-block-format"
	"github.com/ipfs/go-cid"
	"github.com/ipfs/go-ipfs-blockstore"
)

var _ (BlockCache) = (*S3BlockCache)(nil)

type S3BlockCache struct {
	bucket   string
	hc       *http.Client
	upstream BlockCache
	tlogger  logr.Logger // request tracing
	stats    CacheStats
}

func NewS3BlockCache(bucket string, logger logr.Logger) *S3BlockCache {
	if logger == nil {
		logger = logr.Discard()
	}
	return &S3BlockCache{
		bucket:  bucket,
		hc:      &http.Client{},
		tlogger: logger.V(LogLevelTrace),
	}
}

func (s *S3BlockCache) Has(ctx context.Context, c cid.Cid) (bool, error) {
	u := "https://" + path.Join(s.bucket, "mainnet", "blocks", c.String(), "data.raw")
	if s.tlogger.Enabled() {
		s.tlogger.Info("Has", "block", c, "url", u)
	}
	resp, err := s.hc.Head(u)
	if err != nil {
		s.stats.Error()
		if s.tlogger.Enabled() {
			s.tlogger.Error(err, "Has failed", "block", c)
		}
		if s.upstream == nil {
			return false, err
		}
		if s.tlogger.Enabled() {
			s.tlogger.Info("Fulfilling Has from upstream", "block", c)
		}
		return s.upstream.Has(ctx, c)
	}
	if resp.StatusCode == 200 {
		s.stats.Hit()
		return true, nil
	}
	s.stats.Miss()

	if s.tlogger.Enabled() {
		s.tlogger.Info("Block not found", "block", c)
	}
	if s.upstream == nil {
		return false, nil
	}
	if s.tlogger.Enabled() {
		s.tlogger.Info("Fulfilling has from upstream", "block", c)
	}
	return s.upstream.Has(ctx, c)
}

func (s *S3BlockCache) Get(ctx context.Context, c cid.Cid) (blocks.Block, error) {
	u := "https://" + path.Join(s.bucket, "mainnet", "blocks", c.String(), "data.raw")
	if s.tlogger.Enabled() {
		s.tlogger.Info("Get", "block", c, "url", u)
	}
	resp, err := s.hc.Get(u)
	if err != nil {
		s.stats.Error()
		if s.tlogger.Enabled() {
			s.tlogger.Error(err, "Get failed", "block", c)
		}
		if s.upstream == nil {
			return nil, err
		}
		if s.tlogger.Enabled() {
			s.tlogger.Info("Fulfilling get from upstream", "block", c)
		}
		return s.upstream.Get(ctx, c)
	}
	defer resp.Body.Close()
	buf, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode == 200 {
		s.stats.Hit()
		return blocks.NewBlockWithCid(buf, c)
	}
	if s.tlogger.Enabled() {
		s.tlogger.Info("Get failed", "block", c, "http_status", resp.StatusCode)
	}
	s.stats.Miss()

	if s.upstream == nil {
		return nil, blockstore.ErrNotFound
	}

	if s.tlogger.Enabled() {
		s.tlogger.Info("Fulfilling get from upstream", "block", c)
	}
	return s.upstream.Get(ctx, c)
}

func (s *S3BlockCache) SetUpstream(u BlockCache) {
	s.upstream = u
}

func (s *S3BlockCache) LogStats(dlogger logr.Logger) {
	s.stats.Log("s3", dlogger)
}


http.go
package main

import (
	"context"
	"io/ioutil"
	"net/http"
	"strings"

	"github.com/ipfs/go-block-format"
	"github.com/ipfs/go-cid"
	"github.com/ipfs/go-ipfs-blockstore"
)

var _ (BlockCache) = (*HttpBlockCache)(nil)

type HttpBlockCache struct {
	base     string
	hc       *http.Client
	upstream BlockCache
	name     string
}

func NewHttpBlockCache(base string, name string) *HttpBlockCache {
	if !strings.HasSuffix(base, "/") {
		base += "/"
	}

	return &HttpBlockCache{
		base: base,
		name: name,
		hc:   &http.Client{},
	}
}

func (bc *HttpBlockCache) Has(ctx context.Context, c cid.Cid) (bool, error) {
	ctx = cacheContext(ctx, bc.name)
	u := bc.base + c.String() + "/data.raw"
	resp, err := bc.hc.Head(u)
	if err != nil {
		if bc.upstream == nil {
			return false, err
		}
		return bc.upstream.Has(ctx, c)
	}
	if resp.StatusCode == 200 {
		return true, nil
	}

	if bc.upstream == nil {
		return false, nil
	}
	return bc.upstream.Has(ctx, c)
}

func (bc *HttpBlockCache) Get(ctx context.Context, c cid.Cid) (blocks.Block, error) {
	ctx = cacheContext(ctx, bc.name)
	reportEvent(ctx, getRequest)
	stop := startTimer(ctx, getDuration)
	defer stop()

	u := bc.base + c.String() + "/data.raw"
	resp, err := bc.hc.Get(u)
	if err != nil {
		reportEvent(ctx, getFailure)
		if bc.upstream == nil {
			return nil, err
		}
		return bc.upstream.Get(ctx, c)
	}
	defer resp.Body.Close()
	buf, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		reportEvent(ctx, getFailure)
		if bc.upstream == nil {
			return nil, err
		}
		return bc.upstream.Get(ctx, c)
	}
	if resp.StatusCode == 200 {
		reportEvent(ctx, getHit)
		reportSize(ctx, getSize, len(buf))
		return blocks.NewBlockWithCid(buf, c)
	}
	reportEvent(ctx, getMiss)

	if bc.upstream == nil {
		return nil, blockstore.ErrNotFound
	}

	return bc.upstream.Get(ctx, c)
}

func (bc *HttpBlockCache) SetUpstream(u BlockCache) {
	bc.upstream = u
}


node.go
package main

import (
	"context"
	"errors"

	"github.com/go-logr/logr"
	"github.com/ipfs/go-block-format"
	"github.com/ipfs/go-cid"
	"github.com/ipfs/go-ipfs-blockstore"
)

var _ (BlockCache) = (*NodeBlockCache)(nil)

type NodeBlockCacheAPI interface {
	ChainHasObj(ctx context.Context, obj cid.Cid) (bool, error)
	ChainReadObj(ctx context.Context, obj cid.Cid) ([]byte, error)
}

type NodeBlockCache struct {
	node    NodeBlockCacheAPI
	tlogger logr.Logger // request tracing
}

func NewNodeBlockCache(node NodeBlockCacheAPI, logger logr.Logger) *NodeBlockCache {
	if logger == nil {
		logger = logr.Discard()
	}
	return &NodeBlockCache{
		node:    node,
		tlogger: logger.V(LogLevelTrace),
	}
}

func (n *NodeBlockCache) Has(ctx context.Context, c cid.Cid) (bool, error) {
	ctx = cacheContext(ctx, "node")
	has, err := n.node.ChainHasObj(ctx, c)
	if err != nil {
		if errors.Is(err, blockstore.ErrNotFound) {
			return false, err
		}
		if n.tlogger.Enabled() {
			n.tlogger.Error(err, "Has failed", "block", c)
		}
		return false, err
	}

	return has, nil
}

func (n *NodeBlockCache) Get(ctx context.Context, c cid.Cid) (blocks.Block, error) {
	ctx = cacheContext(ctx, "node")
	reportEvent(ctx, getRequest)
	stop := startTimer(ctx, getDuration)
	defer stop()

	data, err := n.node.ChainReadObj(ctx, c)
	if err != nil {
		if errors.Is(err, blockstore.ErrNotFound) {
			reportEvent(ctx, getMiss)
			return nil, err
		}
		reportEvent(ctx, getFailure)
		if n.tlogger.Enabled() {
			n.tlogger.Error(err, "Get failed", "block", c)
		}
		return nil, err
	}

	reportEvent(ctx, getHit)
	reportSize(ctx, getSize, len(data))
	return blocks.NewBlockWithCid(data, c)
}

func (n *NodeBlockCache) SetUpstream(u BlockCache) {
	panic("Not supported")
}


proxy.go
package main

import (
	"context"
	// "fmt"

	"github.com/filecoin-project/go-address"
	"github.com/filecoin-project/go-jsonrpc/auth"
	blocks "github.com/ipfs/go-block-format"
	// "github.com/filecoin-project/go-bitfield"
	"github.com/filecoin-project/go-bitfield"
	"github.com/filecoin-project/go-state-types/abi"
	// "github.com/filecoin-project/go-state-types/dline"
	// "github.com/filecoin-project/go-state-types/network"
	"github.com/filecoin-project/lotus/api"
	"github.com/filecoin-project/lotus/chain/actors/builtin/miner"
	"github.com/filecoin-project/lotus/chain/types"
	"github.com/go-logr/logr"
	"github.com/ipfs/go-cid"
)

type BlockCache interface {
	Has(context.Context, cid.Cid) (bool, error)
	Get(context.Context, cid.Cid) (blocks.Block, error)
	SetUpstream(BlockCache)
}

type Proxy struct {
	node    api.FullNode
	cache   BlockCache
	tlogger logr.Logger // request tracing
}

func NewAPIProxy(node api.FullNode, cache BlockCache, logger logr.Logger) *Proxy {
	if logger == nil {
		logger = logr.Discard()
	}
	return &Proxy{
		node:    node,
		cache:   cache,
		tlogger: logger.V(LogLevelTrace),
	}
}

// Common subset

func (p *Proxy) AuthVerify(ctx context.Context, token string) ([]auth.Permission, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("AuthVerify")
	}
	return p.node.AuthVerify(ctx, token)
}

func (p *Proxy) AuthNew(ctx context.Context, perms []auth.Permission) ([]byte, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("AuthNew")
	}
	return p.node.AuthNew(ctx, perms)
}

func (p *Proxy) Version(ctx context.Context) (api.Version, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("Version")
	}
	return p.node.Version(ctx)
}

// Chain subset

func (p *Proxy) ChainNotify(ctx context.Context) (<-chan []*api.HeadChange, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainNotify")
	}
	return p.node.ChainNotify(ctx)
}

func (p *Proxy) ChainHead(ctx context.Context) (*types.TipSet, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainHead")
	}
	return p.node.ChainHead(ctx)
}

func (p *Proxy) ChainGetBlock(ctx context.Context, obj cid.Cid) (*types.BlockHeader, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainGetBlock", "block", obj)
	}
	sb, err := p.cache.Get(ctx, obj)
	if err != nil {
		if p.tlogger.Enabled() {
			p.tlogger.Error(err, "Failed to get block from cache", "obj", obj)
		}
		return nil, err
	}

	bh, err := types.DecodeBlock(sb.RawData())
	if err != nil {
		if p.tlogger.Enabled() {
			p.tlogger.Error(err, "decode block", "obj", obj, "data", string(sb.RawData()))
		}
	}
	return bh, err
}

func (p *Proxy) ChainGetTipSet(ctx context.Context, tsk types.TipSetKey) (*types.TipSet, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainGetTipSet", "tsk", tsk)
	}
	cids := tsk.Cids()
	blks := make([]*types.BlockHeader, len(cids))
	for i, c := range cids {
		b, err := p.ChainGetBlock(ctx, c)
		if err != nil {
			return nil, err
		}
		blks[i] = b
	}

	ts, err := types.NewTipSet(blks)
	if err != nil {
		return nil, err
	}

	return ts, nil
}

func (p *Proxy) ChainGetBlockMessages(ctx context.Context, blockCid cid.Cid) (*api.BlockMessages, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainGetBlockMessages", "block", blockCid)
	}
	return p.node.ChainGetBlockMessages(ctx, blockCid)
}

func (p *Proxy) ChainGetParentReceipts(ctx context.Context, blockCid cid.Cid) ([]*types.MessageReceipt, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainGetParentReceipts", "block", blockCid)
	}
	return p.node.ChainGetParentReceipts(ctx, blockCid)
}

func (p *Proxy) ChainGetParentMessages(ctx context.Context, blockCid cid.Cid) ([]api.Message, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainGetParentMessages", "block", blockCid)
	}
	return p.node.ChainGetParentMessages(ctx, blockCid)
}

func (p *Proxy) ChainGetTipSetByHeight(ctx context.Context, h abi.ChainEpoch, tsk types.TipSetKey) (*types.TipSet, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainGetTipSetByHeight", "height", h, "tsk", tsk)
	}
	return p.node.ChainGetTipSetByHeight(ctx, h, tsk)
}

func (p *Proxy) ChainReadObj(ctx context.Context, obj cid.Cid) ([]byte, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainReadObj", "obj", obj)
	}
	blk, err := p.cache.Get(ctx, obj)
	if err != nil {
		return p.node.ChainReadObj(ctx, obj)
	}

	return blk.RawData(), nil
}

func (p *Proxy) ChainHasObj(ctx context.Context, obj cid.Cid) (bool, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainHasObj", "obj", obj)
	}
	has, err := p.cache.Has(ctx, obj)
	if err != nil {
		return p.node.ChainHasObj(ctx, obj)
	}
	return has, nil
}

func (p *Proxy) ChainStatObj(ctx context.Context, obj cid.Cid, base cid.Cid) (api.ObjStat, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainStatObj", "obj", obj, "base", base)
	}
	return p.node.ChainStatObj(ctx, obj, base)
}

func (p *Proxy) ChainGetGenesis(ctx context.Context) (*types.TipSet, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainGetGenesis")
	}
	return p.node.ChainGetGenesis(ctx)
}

func (p *Proxy) ChainTipSetWeight(ctx context.Context, tsk types.TipSetKey) (types.BigInt, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainTipSetWeight", "tsk", tsk)
	}
	return p.node.ChainTipSetWeight(ctx, tsk)
}

func (p *Proxy) ChainGetNode(ctx context.Context, path string) (*api.IpldObject, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainGetNode", "path", path)
	}
	return p.node.ChainGetNode(ctx, path)
}

func (p *Proxy) ChainGetMessage(ctx context.Context, mc cid.Cid) (*types.Message, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainGetMessage", "msg", mc)
	}
	return p.node.ChainGetMessage(ctx, mc)
}

func (p *Proxy) ChainGetPath(ctx context.Context, from types.TipSetKey, to types.TipSetKey) ([]*api.HeadChange, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("ChainGetPath", "from", from, "to", to)
	}
	return p.node.ChainGetPath(ctx, from, to)
}

// State subset

func (p *Proxy) StateChangedActors(ctx context.Context, old cid.Cid, new cid.Cid) (map[string]types.Actor, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("StateChangedActors", "old", old, "new", new)
	}
	return p.node.StateChangedActors(ctx, old, new)
}

func (p *Proxy) StateGetReceipt(ctx context.Context, msg cid.Cid, tsk types.TipSetKey) (*types.MessageReceipt, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("StateGetReceipt", "msg", msg, "tsk", tsk)
	}
	return p.node.StateGetReceipt(ctx, msg, tsk)
}

func (p *Proxy) StateListMiners(ctx context.Context, tsk types.TipSetKey) ([]address.Address, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("StateListMiners", "tsk", tsk)
	}
	return p.node.StateListMiners(ctx, tsk)
}

func (p *Proxy) StateListActors(ctx context.Context, tsk types.TipSetKey) ([]address.Address, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("StateListActors", "tsk", tsk)
	}
	return p.node.StateListActors(ctx, tsk)
}

func (p *Proxy) StateGetActor(ctx context.Context, actor address.Address, tsk types.TipSetKey) (*types.Actor, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("StateGetActor", "actor", actor, "tsk", tsk)
	}
	return p.node.StateGetActor(ctx, actor, tsk)
}

func (p *Proxy) StateReadState(ctx context.Context, actor address.Address, tsk types.TipSetKey) (*api.ActorState, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("StateReadState", "actor", actor, "tsk", tsk)
	}
	return p.node.StateReadState(ctx, actor, tsk)
}

func (p *Proxy) StateMinerSectors(ctx context.Context, addr address.Address, sectorNos *bitfield.BitField, tsk types.TipSetKey) ([]*miner.SectorOnChainInfo, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("StateMinerSectors", "addr", addr, "tsk", tsk)
	}
	return p.node.StateMinerSectors(ctx, addr, sectorNos, tsk)
}

func (p *Proxy) StateMinerPower(ctx context.Context, addr address.Address, tsk types.TipSetKey) (*api.MinerPower, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("StateMinerPower", "addr", addr, "tsk", tsk)
	}
	return p.node.StateMinerPower(ctx, addr, tsk)
}

func (p *Proxy) StateVMCirculatingSupplyInternal(ctx context.Context, tsk types.TipSetKey) (api.CirculatingSupply, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("StateVMCirculatingSupplyInternal", "tsk", tsk)
	}
	return p.node.StateVMCirculatingSupplyInternal(ctx, tsk)
}

func (p *Proxy) GetTipSetFromKey(ctx context.Context, tsk types.TipSetKey) (*types.TipSet, error) {
	if p.tlogger.Enabled() {
		p.tlogger.Info("GetTipSetFromKey", "tsk", tsk)
	}
	if tsk.IsEmpty() {
		return p.node.ChainHead(ctx) // equivalent to Chain.GetHeaviestTipSet
	}
	return p.ChainGetTipSet(ctx, tsk)
}


main.go
package main

import (
	"context"
	"errors"
	"fmt"
	"net"
	"net/http"
	"os"
	"os/signal"
	"path/filepath"
	"syscall"
	"time"

	"github.com/filecoin-project/go-jsonrpc"
	"github.com/gorilla/mux"
	"github.com/iand/gonudb"
	"github.com/iand/logfmtr"
	"github.com/urfave/cli/v2"
)

const (
	LogLevelInfo        = 1 // log level increment for informational logging
	LogLevelDiagnostics = 2 // log level increment for diagnostics logging
	LogLevelTrace       = 3 // log level increment for verbose tracing
)

const (
	diagLogInterval         = 5 * time.Minute // interval between logging metrics when diagnostics logging is enabled
	metricReportingInterval = 2 * time.Second // interval between reporting metrics
)

var ErrLotusUnavailable = errors.New("upstream lotus server not available")

func main() {
	app := &cli.App{
		Name:     "lotus-cpr",
		HelpName: "lotus-cpr",
		Usage:    "A caching proxy for Lotus filecoin nodes.",
		Flags: []cli.Flag{
			&cli.IntFlag{
				Name:    "log-level",
				Aliases: []string{"ll"},
				Usage:   fmt.Sprintf("Set verbosity of logs to `LEVEL` (0: off, %d: info, %d:diagnostics, %d:trace).", LogLevelInfo, LogLevelDiagnostics, LogLevelTrace),
				Value:   1,
				EnvVars: []string{"LOTUS_CPR_LOG_LEVEL"},
			},
			&cli.BoolFlag{
				Name:    "humanize-logs",
				Aliases: []string{"hl"},
				Usage:   "Use humanized and colorized log output.",
				Value:   false,
				EnvVars: []string{"LOTUS_CPR_HUMANIZE_LOGS"},
			},
			&cli.StringFlag{
				Name:    "api",
				Usage:   "Multiaddress of Lotus node.",
				EnvVars: []string{"LOTUS_CPR_API"},
				Value:   "/ip4/127.0.0.1/tcp/1234/http",
			},
			&cli.StringFlag{
				Name:     "api-token",
				Usage:    "Read only API token for Lotus node.",
				EnvVars:  []string{"LOTUS_CPR_API_TOKEN"},
				Required: true,
			},
			&cli.StringFlag{
				Name:    "store",
				Usage:   "Path to directory containing gonudb store.",
				EnvVars: []string{"LOTUS_CPR_STORE_PATH"},
			},
			&cli.StringFlag{
				Name:    "blockstore-baseurl",
				Usage:   "Base URL of a web server that serves blocks (urls follow pattern: {blockstore-baseurl}/{block_cid}/data.raw)",
				EnvVars: []string{"LOTUS_CPR_BLOCKSTORE_BASEURL"},
			},
			&cli.StringFlag{
				Name:    "listen",
				Usage:   "Address to start the jsonrpc server on.",
				EnvVars: []string{"LOTUS_CPR_LISTEN"},
				Value:   ":33111",
			},
			&cli.StringFlag{
				Name:    "diag",
				Usage:   "Address to start the diagnostics server on.",
				EnvVars: []string{"LOTUS_CPR_DIAG"},
				Value:   ":33112",
			},
			&cli.IntFlag{
				Name:    "api-concurrency",
				Usage:   "Maximum number of concurrent requests to make to the Lotus node API before triggering disconnection.",
				Value:   2000,
				EnvVars: []string{"LOTUS_CPR_API_CONCURRENCY"},
			},
			&cli.IntFlag{
				Name:    "api-errors",
				Usage:   "Maximum number of errors received from the Lotus node API before triggering disconnection.",
				Value:   8,
				EnvVars: []string{"LOTUS_CPR_API_ERRORS"},
			},
			&cli.DurationFlag{
				Name:    "disconnect-timeout",
				Usage:   "Time to wait after a disconnect from the Lotus node before attempting to reconnect.",
				Value:   30 * time.Second,
				EnvVars: []string{"LOTUS_CPR_DISCONNECT_TIMEOUT"},
			},
		},
		Action:          run,
		HideHelpCommand: true,
	}

	if err := app.Run(os.Args); err != nil {
		fmt.Println(err.Error())
		os.Exit(1)
	}
}

func run(cc *cli.Context) error {
	ctx, cancel := context.WithCancel(cc.Context)
	defer cancel()

	logfmtr.SetVerbosity(cc.Int("log-level"))
	loggerOpts := logfmtr.DefaultOptions()
	if cc.Bool("humanize-logs") {
		loggerOpts.Humanize = true
		loggerOpts.Colorize = true
	}
	logfmtr.UseOptions(loggerOpts)
	logger := logfmtr.New().V(LogLevelInfo)

	// Init metric reporting if required
	reportMetrics := false
	dlogger := logfmtr.New().V(LogLevelDiagnostics)
	if dlogger.Enabled() || cc.String("diag") != "" {
		reportMetrics = true
		if err := initMetricReporting(metricReportingInterval); err != nil {
			return fmt.Errorf("failed to initialize metric reporting: %w", err)
		}
	}

	client, err := newAPIClient(cc.String("api"), cc.String("api-token"), cc.Int("api-errors"), cc.Int("api-concurrency"), cc.Duration("disconnect-timeout"), logfmtr.NewNamed("client"))
	if err != nil {
		return fmt.Errorf("failed to create api client: %w", err)
	}
	defer client.Close()

	caches := []BlockCache{
		NewNodeBlockCache(client, logfmtr.NewNamed("node")),
	}

	if cc.String("blockstore-baseurl") != "" {
		hCache := NewHttpBlockCache(cc.String("blockstore-baseurl"), "http")

		upstream := caches[len(caches)-1]
		hCache.SetUpstream(upstream)

		caches = append(caches, hCache)
		logger.Info("Added http blockstore", "base_url", cc.String("blockstore-baseurl"))
	}

	if cc.String("store") != "" {
		logger.Info("Opening store", "path", cc.String("store"))
		s, err := openStore(ctx, cc.String("store"))
		if err != nil {
			return fmt.Errorf("failed to open gonudb store: %w", err)
		}
		defer func() {
			err := s.Close()
			if err != nil {
				logger.Error(err, "failed to close store cleanly")
			}
		}()

		dbCache := NewDBBlockCache(s, logfmtr.NewNamed("gonudb"))

		if reportMetrics {
			go func() {
				timer := time.NewTicker(2 * time.Second)
				for {
					select {
					case <-timer.C:
						dbCache.ReportMetrics(ctx)
					case <-ctx.Done():
						timer.Stop()
						return
					}
				}
			}()
		}

		upstream := caches[len(caches)-1]
		dbCache.SetUpstream(upstream)

		caches = append(caches, dbCache)
		logger.Info("Added gonudb cache", "path", cc.String("store"))
	}

	rpcServer := jsonrpc.NewServer()
	rpcServer.Register("Filecoin", NewAPIProxy(client, caches[len(caches)-1], logfmtr.NewNamed("proxy")))

	// Set up a signal handler to cancel the context
	go func() {
		interrupt := make(chan os.Signal, 1)
		signal.Notify(interrupt, syscall.SIGTERM, syscall.SIGINT)
		select {
		case <-interrupt:
			cancel()
		case <-ctx.Done():
		}
	}()

	// Log metrics?
	if dlogger.Enabled() {
		go func() {
			timer := time.NewTicker(diagLogInterval)
			ml := NewMetricLogger(dlogger)
			for {
				select {
				case <-timer.C:
					ml.Log()
				case <-ctx.Done():
					timer.Stop()
					return
				}
			}
		}()
	}

	// Serve metrics via http?
	if cc.String("diag") != "" {
		diagListener, err := net.Listen("tcp", cc.String("diag"))
		if err != nil {
			return fmt.Errorf("failed to listen on %q: %w", cc.String("diag"), err)
		}

		pe, err := registerPrometheusExporter("lotuscpr")
		if err != nil {
			return fmt.Errorf("failed to register prometheus exporter: %w", err)
		}

		diagMux := mux.NewRouter()
		diagMux.Handle("/metrics", pe)

		diagSrv := &http.Server{
			Handler: diagMux,
		}

		go func() {
			<-ctx.Done()
			if err := diagSrv.Shutdown(context.Background()); err != nil {
				logger.Error(err, "failed to shut down diagnostics server")
			}
		}()

		logger.Info("Starting diagnostics server", "addr", cc.String("diag"))
		go diagSrv.Serve(diagListener)
	}

	address := cc.String("listen")
	listener, err := net.Listen("tcp", address)
	if err != nil {
		return fmt.Errorf("failed to listen on %q: %w", cc.String("listen"), err)
	}

	mux := mux.NewRouter()
	mux.Handle("/rpc/v0", rpcServer)
	mux.PathPrefix("/").Handler(http.DefaultServeMux)

	srv := &http.Server{
		Handler: mux,
	}

	go func() {
		<-ctx.Done()
		if err := srv.Shutdown(context.Background()); err != nil {
			logger.Error(err, "failed to shut down RPC server")
		}
	}()

	logger.Info("Starting RPC server", "addr", cc.String("listen"))
	return srv.Serve(listener)
}

func openStore(ctx context.Context, path string) (*gonudb.Store, error) {
	datPath := filepath.Join(path, "blocks.dat")
	keyPath := filepath.Join(path, "blocks.key")
	logPath := filepath.Join(path, "blocks.log")

	_, err := os.Stat(datPath)
	if err != nil {
		var pathErr *os.PathError
		if errors.As(err, &pathErr) && os.IsNotExist(pathErr) {
			err := gonudb.CreateStore(
				datPath,
				keyPath,
				logPath,
				1,
				gonudb.NewSalt(),
				4096,
				0.5,
			)
			if err != nil {
				return nil, fmt.Errorf("create store: %w", err)
			}
		} else {
			return nil, fmt.Errorf("stat store: %w", err)
		}
	}

	s, err := gonudb.OpenStore(datPath, keyPath, logPath, &gonudb.StoreOptions{})
	if err != nil {
		return nil, fmt.Errorf("failed to open store: %w", err)
	}
	return s, nil
}
